nixos/tests/stunnel.nix

```diff
@@ -0,0 +1,118 @@
+{ system ? builtins.currentSystem, config ? { }
+, pkgs ? import ../.. { inherit system config; } }:
+
+with import ../lib/testing-python.nix { inherit system pkgs; };
+with pkgs.lib;
+
+let
+  make-cert = { pkgs, ... }: {
+    system.activationScripts.create-test-cert = stringAfter [ "users" ] ''
+      ${pkgs.openssl}/bin/openssl req -batch -x509 -newkey rsa -nodes -out /test-cert.pem -keyout /test-key.pem
+      ( umask 077; cat /test-key.pem /test-cert.pem > /test-key-and-cert.pem )
+      chown stunnel /test-key.pem /test-key-and-cert.pem
+    '';
+    users.groups.stunnel = { };
+    users.users.stunnel = {
+      isSystemUser = true;
+      group = "stunnel";
+    };
+  };
+  server-common = { pkgs, ... }: {
+    networking.firewall.allowedTCPPorts = [ 443 ];
+    services.stunnel = {
+      enable = true;
+      user = "stunnel";
+      servers.https = {
+        accept = "443";
+        connect = 80;
+        cert = "/test-key-and-cert.pem";
+      };
+    };
+    systemd.services.simple-webserver = {
+      wantedBy = [ "multi-user.target" ];
+      script = ''
+        cd /etc/webroot
+        ${pkgs.python3}/bin/python -m http.server 80
+      '';
+    };
+  };
+
+in {
+  basic = makeTest {
+    name = "stunnel";
+
+    nodes = {
+      clint = { };
+      sergey = {
+        imports = [ make-cert server-common ];
+        environment.etc."webroot/index.html".text = "well met";
+      };
+    };
+
+    testScript = ''
+      start_all()
+      sergey.wait_for_unit("simple-webserver")
+      sergey.wait_for_unit("stunnel")
+      clint.succeed("curl --fail --insecure https://sergey/ > out")
+      clint.succeed('[[ "$(< out)" == "well met" ]]')
+    '';
+  };
+
+  mutually-authenticated = makeTest {
+    name = "mutually-authenticated-stunnel";
+
+    nodes = {
+      clint = {
+        imports = [ make-cert ];
+        services.stunnel = {
+          enable = true;
+          user = "stunnel";
+          clients.authenticated-https = {
+            accept = "80";
+            connect = "sergey:443";
+            verifyPeer = true;
+            CAFile = "/authorized-server-cert.crt";
+            extraConfig = ''
+              cert = /test-cert.pem
+              key = /test-key.pem
+            '';
+          };
+        };
+      };
+      sergey = {
+        imports = [ make-cert server-common ];
+        services.stunnel.servers.https.extraConfig = ''
+          CAFile = /authorized-client-certs.crt
+          verifyPeer = yes
+        '';
+        environment.etc."webroot/index.html".text = "secret handshake";
+      };
+    };
+
+    testScript = ''
+      from shlex import quote
+
+      start_all()
+
+      # Introduce server to client
+      sergey.wait_for_file("/test-key-and-cert.pem")
+      server_cert = sergey.succeed("cat /test-cert.pem")
+      clint.succeed("echo %s > /authorized-server-cert.crt" % quote(server_cert))
```

Comment:
(nitpick) might make sense to use the same copy code for first test and have `curl --cafile /authorized-server-cert.crt` instead of `--insecure` ? but to be honest the second tests's client stunnel will verify the server certificate so it doesn't really matter imo.

Reply:
Good idea.  Done.  Thanks!